builtin/logical/pki/storage.go

```diff
@@ -0,0 +1,130 @@
+
+type key struct {
+	ID             keyId                   `json:"id" structs:"id" mapstructure:"id"`
+	PrivateKeyType certutil.PrivateKeyType `json:"private_key_type" structs:"private_key_type" mapstructure:"private_key_type"`
+	PrivateKey     string                  `json:"private_key" structs:"private_key" mapstructure:"private_key"`
+}
+
+type issuer struct {
+	ID           issuerId `json:"id" structs:"id" mapstructure:"id"`
+	Name         string   `json:"name" structs:"name" mapstructure:"name"`
+	KeyID        keyId    `json:"key_id" structs:"key_id" mapstructure:"key_id"`
+	Certificate  string   `json:"certificate" structs:"certificate" mapstructure:"certificate"`
+	CAChain      []string `json:"ca_chain" structs:"ca_chain" mapstructure:"ca_chain"`
+	SerialNumber string   `json:"serial_number" structs:"serial_number" mapstructure:"serial_number"`
+}
+
+func fetchKeyById(ctx context.Context, s logical.Storage, keyId keyId) (key, error) {
+	keyEntry, err := s.Get(ctx, keyPrefix+keyId.String())
+	if err != nil {
+		return emptyKey, errutil.InternalError{Err: fmt.Sprintf("unable to fetch pki key: %v", err)}
+	}
+	if keyEntry == nil {
+		// FIXME: Dedicated/specific error for this?
+		return emptyKey, errutil.UserError{Err: fmt.Sprintf("pki key id %s does not exist", keyId.String())}
+	}
+
+	var key key
+	if err := keyEntry.DecodeJSON(&key); err != nil {
+		return emptyKey, errutil.InternalError{Err: fmt.Sprintf("unable to decode pki key with id %s: %v", keyId.String(), err)}
+	}
+
+	return key, nil
+}
+
+func writeKey(ctx context.Context, s logical.Storage, key key) error {
+	keyId := key.ID
+
+	json, err := logical.StorageEntryJSON(keyPrefix+keyId.String(), key)
+	if err != nil {
+		return err
+	}
+
+	return s.Put(ctx, json)
+}
+
+func fetchIssuerById(ctx context.Context, s logical.Storage, issuerId issuerId) (issuer, error) {
+	issuerEntry, err := s.Get(ctx, issuerPrefix+issuerId.String())
+	if err != nil {
+		return emptyIssuer, errutil.InternalError{Err: fmt.Sprintf("unable to fetch pki issuer: %v", err)}
+	}
+	if issuerEntry == nil {
+		// FIXME: Dedicated/specific error for this?
+		return emptyIssuer, errutil.UserError{Err: fmt.Sprintf("pki issuer id %s does not exist", issuerId.String())}
+	}
+
+	var issuer issuer
+	if err := issuerEntry.DecodeJSON(&issuer); err != nil {
+		return emptyIssuer, errutil.InternalError{Err: fmt.Sprintf("unable to decode pki issuer with id %s: %v", issuerId.String(), err)}
+	}
+
+	return issuer, nil
+}
+
+func writeIssuer(ctx context.Context, s logical.Storage, issuer issuer) error {
+	issuerId := issuer.ID
+
+	json, err := logical.StorageEntryJSON(issuerPrefix+issuerId.String(), issuer)
+	if err != nil {
+		return err
+	}
+
+	return s.Put(ctx, json)
+}
+
+func fetchCertBundleByIssuerId(ctx context.Context, s logical.Storage, issuerId issuerId) (certutil.CertBundle, error) {
+	issuer, err := fetchIssuerById(ctx, s, issuerId)
+	if err != nil {
+		return emptyCertBundle, err
+	}
+
+	if issuer.KeyID == "" {
```

Comment:
Interesting, I had taken the other approach of allowing the bundle to be created, but then (in `fetchCAInfo`) validating the key there.

Reply:
Since so much of the existing code expects a private key to exist within a certbundle that would be a lot of places we would need to add that validation to no?

Reply:
I think we'd only need to add it to `fetchCAInfo` from what I can tell, which is used in the signing path. In the non-signing path (future `/issuer` work), I think we won't need to validate that.

Reply:
Here's what I had for what its worth:

```diff 
+ func fetchIssuerBundle(ctx context.Context, s logical.Storage, id pkiIssuer, loadKey bool) (*certutil.CertBundle, error) {
+       var bundle certutil.CertBundle
+
+       issuer, err := fetchPKIIssuerById(ctx, s, id)
+       if err != nil {
+               return nil, err
+       }
+
+       bundle.Certificate = issuer.Certificate
+       bundle.IssuingCA = issuer.IssuingCA
+       bundle.CAChain = issuer.CAChain
+       bundle.SerialNumber = issuer.SerialNumber
+
+       // Fetch the key if it exists. Sometimes we don't need the key immediately.
+       var key pkiKey
+       if loadKey && issuer.PKIKeyID != pkiKeyId("") {
+               key, err = fetchPKIKeyById(ctx, s, id)
+               if err != nil {
+                       return nil, err
+               }
+
+               bundle.PrivateKeyType = key.PrivateKeyType
+               bundle.PrivateKey = key.PrivateKey
+       }
+
+       return &bundle, nil
+}
```

I thought returning pointer was also cleaner. My 2c :-)

Reply:
How about two methods instead of the boolean (it's a bit of a pet peeve and harder to identify who does what). As for the pointers sure I can switch as I suppose that is the standard in Vault.

Reply:
Thinking about it a little more let's go with your implementation. I'll remove mine from my PR for now.

Reply:
I've added it to my PR: https://github.com/hashicorp/vault/pull/14804